core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -86,14 +89,64 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
-   * @throws KeeperException if an error is returned by ZooKeeper.
+   * @throws ControllerMovedException if fail to create /controller or fail to increment controller epoch.
    */
-  def registerController(controllerId: Int, timestamp: Long): Unit = {
-    val path = ControllerZNode.path
-    checkedEphemeralCreate(path, ControllerZNode.encode(controllerId, timestamp))
+  def registerControllerAndIncrementControllerEpoch(controllerId: Int): (Int, Int) = {
+    val timestamp = time.milliseconds()
+
+    // Create /controller_epoch if not exists
+    maybeCreateControllerEpochZNode()
+
+    // Read /controller_epoch to get the current controller epoch and zkVersion
+    val (curEpoch, curEpochStat) = getControllerEpoch.getOrElse(
+      throw new IllegalStateException(s"${ControllerEpochZNode.path} does not exist while trying to create ${ControllerZNode.path}"))
+
+    // Create /controller and update /controller_epoch atomically
+    val newControllerEpoch = curEpoch + 1
+    val expectedControllerEpochZkVersion = curEpochStat.getVersion
+
+    debug(s"Try to create ${ControllerZNode.path} and increment controller epoch to $newControllerEpoch with expected controller epoch zkVersion $expectedControllerEpochZkVersion")
+    try {
+      val transaction = zooKeeperClient.createTransaction()
+      transaction.check(ControllerEpochZNode.path, expectedControllerEpochZkVersion)
+      transaction.create(ControllerZNode.path, ControllerZNode.encode(controllerId, timestamp),
+        acls(ControllerZNode.path).asJava, CreateMode.EPHEMERAL)
+      transaction.setData(ControllerEpochZNode.path, ControllerEpochZNode.encode(newControllerEpoch), expectedControllerEpochZkVersion)
+      val results = transaction.commit()
+      val setDataResult = results.get(2).asInstanceOf[SetDataResult]
+      (newControllerEpoch, setDataResult.getStat.getVersion)
+    } catch {
+      case _: NodeExistsException =>
+        val curControllerId = getControllerId.getOrElse(throw new ControllerMovedException(
+          s"The ephemeral node at ${ControllerZNode.path} went away while checking whether the controller election succeeds. " +
+          s"Aborting controller startup procedure"))
+        if (controllerId == curControllerId) {
+          val (epoch, stat)  = getControllerEpoch.getOrElse(throw new ControllerMovedException(
+            "Controller moved to another broker. Aborting controller startup procedure"))
+          // If the epoch is the same as newControllerEpoch, it is safe to infer that the returned epoch zkVersion
+          // is associated with the current broker during controller election because we already knew that the zk
+          // transaction succeeds based on the controller znode verification. Other rounds of controller
+          // election will result in larger epoch number written in zk.
+          if (epoch == newControllerEpoch)
+            (newControllerEpoch, stat.getVersion)
```

Comment:
Should we explicitly call return here? Otherwise, it seems that we are always throwing ControllerMovedException.

Reply:
Ah... My bad. Fixed.